src/plugins/tonegen/rfc4733.c

```diff
@@ -0,0 +1,424 @@
+        sts = dbusif_register_input_method(tonegend, m->intf, m->memb,
+                                           m->sig, m->func);
+
+        if (sts < 0) {
+            N_ERROR(LOG_CAT "%s(): Can't register D-Bus method '%s'",
+                      __FUNCTION__, m->memb);
+            err = -1;
+        }
+    }
+
+    return err;
+}
+
+int rfc4733_create_ngfd(struct tonegend *tonegend)
+{
+    struct method_ngfd *m;
+    int sts;
+    int err = 0;
+
+    for (m = method_ngfd_defs, err = 0; m->name; m++) {
+        sts = ngfif_register_input_method(tonegend, m->name, m->func_start, m->func_stop);
+        if (sts < 0) {
+            N_ERROR(LOG_CAT "Failed to register method %s", m->name);
+            err = -1;
+        }
+    }
+
+    indicator_hash = g_hash_table_new_full(g_str_hash, g_str_equal,
+                                           g_free, NULL);
+
+    g_hash_table_insert(indicator_hash, g_strdup("dial"),       GUINT_TO_POINTER(TONE_DIAL));
+    g_hash_table_insert(indicator_hash, g_strdup("busy"),       GUINT_TO_POINTER(TONE_BUSY));
+    g_hash_table_insert(indicator_hash, g_strdup("congest"),    GUINT_TO_POINTER(TONE_CONGEST));
+    g_hash_table_insert(indicator_hash, g_strdup("radio_ack"),  GUINT_TO_POINTER(TONE_RADIO_ACK));
+    g_hash_table_insert(indicator_hash, g_strdup("radio_na"),   GUINT_TO_POINTER(TONE_RADIO_NA));
+    g_hash_table_insert(indicator_hash, g_strdup("error"),      GUINT_TO_POINTER(TONE_ERROR));
+    g_hash_table_insert(indicator_hash, g_strdup("wait"),       GUINT_TO_POINTER(TONE_WAIT));
+    g_hash_table_insert(indicator_hash, g_strdup("ring"),       GUINT_TO_POINTER(TONE_RING));
+
+    return err;
+}
+
+void rfc4733_destroy()
+{
+    if (indicator_hash) {
+        g_hash_table_destroy(indicator_hash);
+        indicator_hash = NULL;
+    }
+}
+
+static int start_indicator_tone_ngfd(NRequest *request, struct tonegend *tonegend)
+{
+    struct ausrv *ausrv = tonegend->ausrv_ctx;
+    gpointer      event_p;
+    uint32_t      event;
+    int32_t       dbm0 = 0;
+    uint32_t      duration = 0;
+    uint32_t      volume;
+    const NProplist *proplist;
+    const NValue *value;
+
+    proplist = n_request_get_properties(request);
+    N_DEBUG(LOG_CAT "request indicator event");
+
+    if (!n_proplist_has_key(proplist, "tonegen.pattern")) {
+        N_WARNING(LOG_CAT "request doesn't have pattern.");
+        return FALSE;
+    }
+
+    value = n_proplist_get(proplist, "tonegen.pattern");
+    if (n_value_type(value) == N_VALUE_TYPE_STRING) {
+
+        if (!(event_p = g_hash_table_lookup(indicator_hash,
+                                            n_value_get_string(value)))) {
+            N_WARNING(LOG_CAT "request doesn't have valid pattern.");
+            return FALSE;
+        }
+        event = GPOINTER_TO_UINT(event_p);
+
+    } else if (n_value_type(value) == N_VALUE_TYPE_UINT) {
+
+        switch (n_value_get_uint(value)) {
+        case 66:    event = TONE_DIAL;      break;
```

Comment:
(or reference the page of the relevant standard/spec if there is one that defines those values literally)